x/leverage/README.md

```diff
@@ -154,52 +157,83 @@ Supply utilization of a token denomination is calculated as:
+For example, an account with using a single collateral token with `CollateralWeight 0.8` borrowing an single token with `CollateralWeight 0.7` will reduce the effective `CollateralWeight` of the account's collateral to `0.7` when computing borrow limit.
 
-- Borrowed value must be less than collateral value times `CollateralWeight` (sum over each collateral asset)
-- Borrowed value times `BorrowFactor` (sum over each borrowed asset) must be less than collateral value.
+#### Special Asset Pairs
 
-This means that when the original borrow limit based on collateral weight would allow a higher quality collateral to borrow a risky asset with a small margin of safety, the user's effective collateral weight is reduced to that of the riskier asset.
-(Or `0.5` at the minimum.)
+The leverage module can define pairs of assets which are advantaged when one is used as collateral to borrow the other.
 
-#### Historic Borrow Limit, Value
+They are defined in the form `[Asset A, Asset B, Special Collateral Weight]`. In effect, this means that
 
-The leverage module also makes use of the oracle's historic prices to enforce an additional restriction on borrowing.
+> When collateral of `Asset A` exists on the same account as borrowed `Asset B`, the `CollateralWeight` of `Asset A` is replaced by `Special Collateral Weight` when computing collateral weights, and the `CollateralWeight` of `Asset B` is replaced by `Special Collateral Weight` when computing `Borrow Factor`.
 
-The logic is:
+#### Borrow Limit
 
-- For any `MsgBorrow`, `MsgMaxBorrow`, `MsgDecollateralize`, `MsgWithdraw`, or `MsgMaxWithdraw`
-- The borrower’s borrowed value must be less than their borrow limit, with borrowed value being computed using `PriceModeHigh`, i.e. the higher of either spot price or historic price is used.
-- Where historic prices are defined as the Median of the last `N` historic medians from the `oracle` module with `N = Token.HistoricMedians` in the leverage registry
-- Else the transaction fails
+A user's borrow limit is the sum of the contributions from each denomination of collateral they have deposited, with some modifications due to `Borrow Factor` and `Special Asset Pairs`.
+
+The full calculation of a user's borrow limit is as follows:
+
+1. Calculate the USD value of the user's collateral assets, using the _lower_ of either spot price or historic price for each asset. Collateral with missing prices is treated as zero-valued.
+2. Calculate the USD value of the user's borrowed assets, using the _higher_ of either spot price or historic price for each asset. Borrowed assets with missing prices cause any transaction which could increased borrowed value or decrease borrow limit to fail.
+3. If the part or all of a user's position matches any `Special Asset Pairs`, subtract them from the general totals above and set them aside. A borrowed or collateral asset matching multiple special pairs attempts to use the one with the highest `Special Collateral Weight` first.
```

Comment:
1. `[10A] -> [7A]` (looping) operates under no special asset pairs, so it is unchanged compared to current leverage module.
2. `[10A, 10C] -> [7B, 7D]` with `CW_ab = 0.9` behaves as `[7.77A] -> [7B]` + `[2.23A, 10C] -> [7D]` - that is, the special pair is extracted out and the rest of the collateral weights are interpreted normally.
3. `[10A, 10C] -> [15B]` with `CW_ab = 0.9` behaves as `[10A] -> [9B]` + `[10C] -> [6B]`

In each case, the special asset pair is consumes as much amount as it can (while matching collateral to borrows) and any remainder is interpreted as regular position.

The approach of taking a dollar-value-weighted-average of collateral weight after special pairs have been subtracted from base assets, then doing both collateral weight and borrow factor calculations, produces these results.

Reply:
The steps have now been redone - no longer are weighted average mentioned (except in liquidation threshold), and a greedy matching algorithm is described instead. This matches implementation.

Reply:
Added section Special Asset Pair Examples